cirq-core/cirq/devices/insertion_noise_model.py

```diff
@@ -0,0 +1,75 @@
+# Copyright 2021 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from dataclasses import dataclass, field
+from typing import TYPE_CHECKING, Dict, List, Sequence
+
+from cirq import devices, ops
+from cirq.devices.noise_utils import (
+    OpIdentifier,
+    PHYSICAL_GATE_TAG,
+)
+
+if TYPE_CHECKING:
+    import cirq
+
+
+@dataclass
+class InsertionNoiseModel(devices.NoiseModel):
+    """Simple base noise model for inserting operations.
+
+    Operations generated by this model for a given moment are all added into a
+    single "noise moment", which is added before or after the original moment
+    based on `prepend`.
+
+    Args:
+        ops_added: a map of gate types (and optionally, qubits they act on) to
+            operations that should be added.
+        prepend: whether to add the new moment before the current one.
+        require_physical_tag: whether to only apply noise to operations tagged
+            with PHYSICAL_GATE_TAG.
+    """
+
+    ops_added: Dict[OpIdentifier, 'cirq.Operation'] = field(default_factory=dict)
+    prepend: bool = False
+    require_physical_tag: bool = True
+
+    def noisy_moment(
+        self, moment: 'cirq.Moment', system_qubits: Sequence['cirq.Qid']
+    ) -> 'cirq.OP_TREE':
+        noise_ops: List['cirq.Operation'] = []
+        for op in moment:
+            if self.require_physical_tag and PHYSICAL_GATE_TAG not in op.tags:
+                # Only non-virtual gates get noise applied.
+                continue
+            op_id = OpIdentifier(type(op.gate), *op.qubits)
+            if op_id in self.ops_added:
+                noise_ops.append(self.ops_added[op_id])
+                continue
+            # Find the closest match, if one exists.
+            parent_id = OpIdentifier(object, *op.qubits)
+            for added_id in self.ops_added:
+                if added_id.qubits != parent_id.qubits:
+                    continue
+                if not issubclass(op_id.gate_type, added_id.gate_type):
+                    continue
+                if issubclass(added_id.gate_type, parent_id.gate_type):
+                    parent_id = added_id
+            if parent_id.gate_type != object:
+                noise_ops.append(self.ops_added[parent_id])
+        if not noise_ops:
+            return [moment]
+        if self.prepend:
+            return [ops.Moment(noise_ops), moment]
+        return [moment, ops.Moment(noise_ops)]
```

Comment:
This is pretty hard to read. Also, the object-like containment check that falls back to type-like containment check is something gatesets can do for you. Also if there are gates or operations in `self.ops_added` with overlapping parent types in `self.ops_added` this will have strange behavior. Could we tighten this up a bit ?
Maybe something along the lines of:
```python3
non_virtual_ops = [op if self.require_physical_tag and PHYSICAL_GATE_TAG not in op.tags for op in moment]
mappings = self.ops_added.items()
converted_ops = [v(*op.qubits) for k,v in mappings for op in non_virtual_ops if op in k]
return converted_ops + moment
```
where `self.ops_added` maps from gatesets to operations ?

Reply:
`cirq.Gateset` is not serializable, which causes complications when interacting with `NoiseProperties`. It's likely possible to get this working, but I expect it to be more complicated than the current behavior.

RE overlapping parent types: this is addressed here and covered in tests. If you have something like:
```
class A: {...}
class B(A): {...}
class C(B): {...}

ops_added = {
    OpIdentifier(A, q0): cirq.X(q0),
    OpIdentifier(B, q0): cirq.Z(q0),
}
```
then objects of type B or C will generate `cirq.Z(q0)` as noise, and objects of type A will produce `cirq.X(q0)` as noise.

Reply:
> `cirq.Gateset` is not serializable, which causes complications when interacting with `NoiseProperties`

Can this be fixed easily @tanujkhattar ?

Reply:
Discussed offline. @95-martin-orion will try to replace the existing logic with gate families and gatesets. I will add 2 new features to gatesets 
- A method that can return the matched gate family instead of just returning true/false while checking for containment
- If all gate families of a Gateset are serializable, then the gateset should be serializable.

Reply:
Discussed out-of-band with @tanujkhattar - I can reconstruct this to use `Gateset`s and `GateFamily`s with a fairly low attached cost. I'll probably need to rehash these PRs again to support that, but until then I'm happy to take any other comments you have, since the remaining structure of this change should remain similar.

Reply:
Changes have been applied to #4671, and will be added here soon. PTAL